### PR TITLE
Consistent way of not instrumenting multiple times

### PR DESCRIPTION
**Pull request by lzchen**
Fixes https://github.com/open-telemetry/opentelemetry-python-contrib/issues/439
We use `_is_instrumented_by_opentelemetry` for all functionality involving indicating whether a component has been instrumented. This is the text that should be used for any context variable/property that is used revolving around determining whether an instrumentation is already instrumented or not.

Another thing to note is the relationship between `instrument()` and `instrument_app`/`instrument_connection()` etc.
`instrument_X()` should take priority over `instrument()`, i.e. if `uninstrument_app(app)` is called after `instrument()`, `app` should produce no telemetry.

As well, calling `instrument_app(app)` after `instrument()` should not produce double telemetry.

Also a bunch of other consistency-based changes:

1. Made all `instrument/uninstrument_X` static methods, where `X` is a component
2. fastapi - added `uninstrument_app`
3. Changed `warning` to `error` on dependency conflict for `instrument()`

### CODE REVIEW

**Review by codeboten — COMMENTED** [review]
> We use the string `_is_instrumented_by_opentelemetry` for all functionality involving indicating whether a component has been instrumented.

@lzchen this statement in the PR description was confusing to me, that variable is a boolean, not a string correct?

**Comment by lzchen** [comment]
@codeboten 
Yes it is a boolean. In the description I was outlining that we are using this specific text`_is_instrumented_by_opentelemetry` (instead of some other text like `is_instrumented`) for all things related to instrumenting. I've updated my comment.

File: instrumentation/opentelemetry-instrumentation-pyramid/src/opentelemetry/instrumentation/pyramid/__init__.py [hunk 1]
```diff
@@ -79,6 +79,7 @@
 """
 
 import typing
+from os import stat
```
Comment (lonewolf3739) [inline instrumentation/opentelemetry-instrumentation-pyramid/src/opentelemetry/instrumentation/pyramid/__init__.py:82]: nit: unused import

File: instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py [hunk 2]
```diff
@@ -250,28 +254,22 @@ def instrument_app(
...
+        if hasattr(app, "_original_wsgi_app"):
             app.wsgi_app = app._original_wsgi_app
 
             # FIXME add support for other Flask blueprints that are not None
             app.before_request_funcs[None].remove(app._before_request)
             app.teardown_request_funcs[None].remove(_teardown_request)
             del app._original_wsgi_app
-
-            app._is_instrumented = False
+            app._is_instrumented_by_opentelemetry = False
```
Comment (lonewolf3739) [inline instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py:272]: Should we rather delete this attribute when uninstrumented?

**Comment by codeboten** [comment]
Should this be specified in the spec or is this implementation specific details? https://github.com/open-telemetry/oteps/pull/165/files?short_path=fac8c59#diff-fac8c5914f3523b6888b428ec9eea8229238297768d93e2171418f7c45f805ce

File: instrumentation/opentelemetry-instrumentation-aiopg/src/opentelemetry/instrumentation/aiopg/wrappers.py [hunk 3]
```diff
@@ -153,6 +153,10 @@ def instrument_connection(
     Returns:
         An instrumented connection.
     """
+    if isinstance(connection, wrapt.ObjectProxy):
```
[the remark this answers is not in the document]
Comment (lzchen) [reply instrumentation/opentelemetry-instrumentation-aiopg/src/opentelemetry/instrumentation/aiopg/wrappers.py:156]: Yeah I was thinking about that too (differentiating OpenTelemetry proxy objects). We can actually just check for [TracedConnectionProxy](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-aiopg/src/opentelemetry/instrumentation/aiopg/aiopg_integration.py#L60) instead. Thoughts?

File: instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py (hunk 2 above)
Comment (lzchen) [reply instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py:272]: Not a strong opinion on this. The field is really only used by OpenTelemetry.

**Review by lonewolf3739 — APPROVED** [review]

**Review by codeboten — APPROVED** [review]
👍 Thanks for addressing my comments

**Pull request merged**